astropy/uncertainty/core.py

```diff
@@ -49,16 +49,17 @@ def __new__(cls, samples):
         new_dtype = np.dtype({'names': ['samples'],
                               'formats': [(samples.dtype, (samples.shape[-1],))]})
         samples_cls = type(samples)
-        if not issubclass(samples_cls, Distribution):
-            # Make sure first letter is uppercase, but note that we can't use
-            # str.capitalize since that converts the rest of the name to lowercase.
-            new_name = samples_cls.__name__[0].upper() + samples_cls.__name__[1:] + cls.__name__
-            if new_name in cls._generated_subclasses:
-                new_cls = cls._generated_subclasses[new_name]
-            else:
-                new_cls = type(new_name, (cls, samples_cls),
-                               {'_samples_cls': samples_cls})
-                cls._generated_subclasses[new_name] = new_cls
+        new_cls = cls._generated_subclasses.get(samples_cls)
+        if new_cls is None:
+            # Make a new class with the combined name, using the highest
+            # base if it is already available.  E.g., if there were a
+            # AngleDistribution with special properties, LongitudeDistribution
+            # would be based on it.
+            new_name = samples_cls.__name__ + cls.__name__
+            new_cls = type(new_name, (samples_cls, NdarrayDistribution),
```

Comment:
Here is the reversed order (except for `ndarray`, which is defined explicitly below, since I needed the scalar too)